From 504c86dbdcbf2623d4a1aabbf575464cbdd584f3 Mon Sep 17 00:00:00 2001
From: Christopher Egert <cme3000@gmail.com>
Date: Sun, 21 Jun 2020 18:59:24 +0200
Subject: [PATCH] [d3d9] Only force AF on textures with mipmaps.

This fixes some visual artifacts on nvidia with UE3 when
anisotropic filtering is enabled through d3d9.samplerAnisotropy.
---
 src/d3d9/d3d9_device.cpp | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/src/d3d9/d3d9_device.cpp b/src/d3d9/d3d9_device.cpp
index cabf79fb3..121eb2187 100644
--- a/src/d3d9/d3d9_device.cpp
+++ b/src/d3d9/d3d9_device.cpp
@@ -3583,6 +3583,9 @@ namespace dxvk {
 
     UpdateActiveTextures(StateSampler, combinedUsage);
 
+    if (m_d3d9Options.samplerAnisotropy != -1)
+      m_dirtySamplerStates |= 1u << StateSampler;
+
     return D3D_OK;
   }
 
@@ -5480,11 +5483,17 @@ namespace dxvk {
     key.BorderColor   = D3DCOLOR(state[D3DSAMP_BORDERCOLOR]);
 
     if (m_d3d9Options.samplerAnisotropy != -1) {
-      if (key.MagFilter == D3DTEXF_LINEAR)
-        key.MagFilter = D3DTEXF_ANISOTROPIC;
+      DWORD level = 0;
+      if (m_state.textures[Sampler])
+        level = m_state.textures[Sampler]->GetLevelCount();
+
+      if (level > 1) {
+        if (key.MagFilter == D3DTEXF_LINEAR)
+          key.MagFilter = D3DTEXF_ANISOTROPIC;
 
-      if (key.MinFilter == D3DTEXF_LINEAR)
-        key.MinFilter = D3DTEXF_ANISOTROPIC;
+        if (key.MinFilter == D3DTEXF_LINEAR)
+          key.MinFilter = D3DTEXF_ANISOTROPIC;
+      }
 
       key.MaxAnisotropy = m_d3d9Options.samplerAnisotropy;
     }
